README.md
# FishRAM

FishRAM (Fisheries Resource Allocation Model) is an R package to simulate biological and economic processes of commercial and recreational fisheries. The model was initially developed by Tidbury et al. (2021) for use in Sea Bass fisheries, however the tools presented here are general purpose and can be applied to any fishery where simultaneous management of commercial and recreational fleets is required.

## Installation
You can install the development version of FishRAM using the `devtools` package:
``` r
library(devtools)
install_github("CefasRepRes/FishRAM")
```

## Example

To get started with `FishRAM`, the `"seabass"` dataset contained in the package contains the `BioeconomicParams` object encapsulating the parameters for the Sea Bass fishery studied in Tidbury et al. (2021). This can be loaded in and a simulation run as follows:

``` r
library(FishRAM)
#Load in the params for Sea Bass in the North Sea
data("seabass")
#Change the adult mortality parameter by changing the slot in params.
params@muA <- 0.1
# Specify recruitment as 15 years of high recruitment followed by 15 years of low recruitment
recruitment <- c(rep(23151200, 15), rep(757600, 15))
#Run a simulation
sim <- project(params, R = recruitment, t_start = 1991, t_end = 2020)
```
The `project()` function returns a `BioeconomicSim` object encapsulating the model results. We can call plotting functions on this object to explore the various predictions of the model.

The `plotStock()` function gives the stock sizes of the adult and juvenile stocks over the duration of the simulation.
``` r
plotStock(sim)
```
![](man/figures/stockplot.png)

The `plotActivity()` function plots the number of trips each year for the commercial or recreational fleets. 
``` r
plotActivity(sim, fleet = "C") # The commercial fleet activity
plotActivity(sim, fleet = "R") # The recreational fleet activity
```
![](man/figures/activity_com.png)
![](man/figures/activity_rec.png)

Finally, the `plotEconomicImpact()` function plots the GVA of each fleet for each year.

``` r
plotEconomicImpact(sim, fleet = "T") # Plot the total GVA across both fleets.
```
![](man/figures/impact_plot.png)


# References
Tidbury, Hannah J, Angela Muench, Philip D Lamb, and Kieran Hyder. 2021. “Balancing Biological and Economic Goals in Commercial and Recreational Fisheries: Systems Modelling of Sea Bass Fisheries.” *ICES Journal of Marine Science* 78: 1793–1803.


NEWS.md
# FishRAM 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/FiRAM-package.R
#' Fisheries Resource Allocation Model
#'
#' @description The `FiRAM` package implements the model from Tidbury et. al. (2021) to simulate biological and economic impacts of recreational and commercial fleets.
#'
#' @details
#' Running simulations in `FiRAM` is done in 3 main stages:
#' 1. Set up model parameters using the `BioeconomicParams` constructor function.
#' 2. Run simulations using the `project` function and a recruitment scenario. This creates a `BioeconomicSim` object.
#' 3. Explore results by calling plotting functions (`plot, plotStock, plotActivity, plotEconomicImpact`) on the `BioeconomicSim` object.
#'
#' @docType package
#' @name FiRAM-package
#' @aliases FiRAM
#' @references Hannah J Tidbury, Angela Muench, Philip D Lamb, Kieran Hyder, "Balancing biological and economic goals in commercial and recreational fisheries: systems modelling of sea bass fisheries", *ICES Journal of Marine Science*, Volume 78, Issue 5, August 2021, Pages 1793-1803
NULL


R/plot_functions.R
#' Plot the stock for a simulation
#' @param sim A `BioEconomicSim` object.
#' @param stock A `character` giving the stock desired. This should be `"A"` for the adult population, `"J"` for the juvenile population, `"T"` for the total population  or `"B"` for both stocks plotted on the same axis.
#' @examples
#' #load in the model parameters
#' data("seabass")
#' R <- c(rep(23151200, 20), rep(757600, 20))
#' sim <- project(params, R = R)
#' plotStock(sim)
#' plotStock(sim, stock = "T")
#' plotStock(sim, stock = "A")
#' @export
plotStock <- function(sim, stock = "B"){
  if(!is(sim, "BioeconomicSim")){
    stop("sim should be ")
  }

  if(!(stock %in% c("B", "T", "J", "A"))){
    warning("Stock should be a character specifying the stock type that should be plotted. Plotting both stocks.")
  }

  years <- sim@t_start:sim@t_end
  if (stock == "A"){
    df <- cbind(years, sim@states[, "SA"])
    colnames(df) <- c("Year", "Adult popn (000)s")
    df[, "Adult popn (000)s"] <- df[, "Adult popn (000)s"] / 1000
    df <- as.data.frame(df)
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Adult popn (000)s`)) + ggplot2::geom_line()
    )
  }else if(stock == "J"){
    df <- cbind(years, sim@states[, "SJ"])
    colnames(df) <- c("Year", "Juvenile popn (000)s")
    df[, "Juvenile popn (000)s"] <- df[, "Juvenile popn (000)s"] / 1000
    df <- as.data.frame(df)
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Juvenile popn (000)s`)) + ggplot2::geom_line()
    )
  }else if(stock == "T"){
    df <- data.frame(years, rowSums(sim@states[, c("SA", "SJ")]))
    colnames(df) <- c("Year", "Population (000s)")
    df[, "Population (000s)"] <- df[, "Population (000s)"] / 1000
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Population (000s)`)) + ggplot2::geom_line()
    )

  }
  df <- cbind(years, sim@states[, c("SJ", "SA")])
  colnames(df) <- c("Year", "Juvenile", "Adult")
  df[, "Juvenile"] <- df[, "Juvenile"] / 1000
  df[, "Adult"] <- df[, "Adult"] / 1000

  df <- reshape2::melt(df, id.vars="Year")
  colnames(df) <- c("Year", "Stock", "Population (000s)")
   return(
    ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Population (000s)`, group = `Stock`, colour = `Stock`)) + ggplot2::geom_line()
  )
}

#' Plot the economic impact for a simulation
#' @param sim A `BioEconomicSim` object.
#' @param fleet A `character` specifying which fleet to plot. This should be one of `"C"` to plot the commercial fleet, `"R"` to plot the recreational fleet, "T" to plot the total GVA, or `"B"` to plot both fleets on the same axis.
#' @examples
#' #load in the model parameters
#' data("seabass")
#' R <- c(rep(23151200, 20), rep(757600, 20))
#' sim <- project(params, R = R)
#' plotEconomicImpact(sim)
#' plotEconomicImpact(sim, fleet = "T")
#' plotEconomicImpact(sim, fleet = "C")
#'
#' @export
plotEconomicImpact <- function(sim, fleet = "B"){
  if(!(fleet %in% c("B", "R", "T", "C")))
    warning("Fleet should be one of R, C, T, or B to specify the fleet. Plotting both fleets.")


  years <- sim@t_start:sim@t_end
  if(fleet == "C"){
    df <- data.frame(years, sim@states[, "VC"])
    colnames(df) <- c("Year", "Commercial GVA")
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Commercial GVA`)) + ggplot2::geom_line()
    )
  }else if(fleet == "R"){
    df <- cbind(years, sim@states[, "VR"])
    colnames(df) <- c("Year", "Recreational GVA")
    df <- as.data.frame(df)
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Recreational GVA`)) + ggplot2::geom_line()
    )
  }
  df <- data.frame(years, sim@states[, c("VR", "VC")])
  colnames(df) <- c("Year", "Recreational GVA", "Commercial GVA")
  df <- reshape2::melt(df, id.vars = "Year")
  colnames(df) <- c("Year", "Fleet", "GVA")
  if(fleet == "T"){
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `GVA`, fill = `Fleet`)) + ggplot2::geom_bar(stat = "identity")
    )
  }
  return(
    ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `GVA`, group = `Fleet`, colour = `Fleet`)) + ggplot2::geom_line()
  )
}


#' Plot the activity (number of trips) for a simulation
#' @param sim A `BioEconomicSim` object.
#' @param fleet A `character` specifying which fleet to plot. This should be one of `"C"` to plot the commercial fleet, `"R"` to plot the recreational fleet, or `"B"` to plot both fleets on the same axis.
#' @examples
#' #load in the model parameters
#' data("seabass")
#' R <- c(rep(23151200, 20), rep(757600, 20))
#' sim <- project(params, R = R)
#' plotActivity(sim)
#' plotActivity(sim, fleet = "R")
#' plotActivity(sim, fleet = "C")
#' @export
plotActivity <- function(sim, fleet = "B"){
  if(!(fleet %in% c("B", "R", "C")))
    warning("Fleet should be one of R, C, or B to specify the fleet. Plotting both fleets.")


  years <- sim@t_start:sim@t_end
  if(fleet == "C"){
    df <- data.frame(years, sim@states[, "TC"])
    df[, 2] <- df[, 2] / 1000
    colnames(df) <- c("Year", "Commercial Trips (000s)")
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Commercial Trips (000s)`)) + ggplot2::geom_line()
    )
  }else if(fleet == "R"){
    df <- cbind(years, sim@states[, "TR"])
    df[, 2] <- df[, 2] / 1000
    colnames(df) <- c("Year", "Recreational Trips (000s)")
    df <- as.data.frame(df)
    return(
      ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Recreational Trips (000s)`)) + ggplot2::geom_line()
    )
  }
  df <- data.frame(years, sim@states[, c("TR", "TC")])
  df[, c("TR", "TC")] <- df[, c("TR", "TC")] / 1000

  colnames(df) <- c("Year", "Recreational Trips (000s)","Commercial Trips (000s)")
  df <- reshape2::melt(df, id.vars = "Year")
  colnames(df) <- c("Year", "Fleet", "Trips (000s)")
  return(ggplot2::ggplot(df, ggplot2::aes(x=`Year`, y = `Trips (000s)`, group = `Fleet`, colour = `Fleet`)) + ggplot2::geom_line())
}



R/FishRAM-package.R
#' Fisheries Resource Allocation Model
#'
#' @description The `FishRAM` package implements the model from Tidbury et. al. (2021) to simulate biological and economic impacts of recreational and commercial fleets.
#'
#' @details
#' Running simulations in `FishRAM` is done in 3 main stages:
#' 1. Set up model parameters using the `BioeconomicParams` constructor function.
#' 2. Run simulations using the `project` function and a recruitment scenario. This creates a `BioeconomicSim` object.
#' 3. Explore results by calling plotting functions (`plot, plotStock, plotActivity, plotEconomicImpact`) on the `BioeconomicSim` object.
#'
#' @docType package
#' @name FishRAM-package
#' @aliases FishRAM
#' @references Hannah J Tidbury, Angela Muench, Philip D Lamb, Kieran Hyder, "Balancing biological and economic goals in commercial and recreational fisheries: systems modelling of sea bass fisheries", *ICES Journal of Marine Science*, Volume 78, Issue 5, August 2021, Pages 1793-1803
NULL


R/BioeconomicSim-class.R
#' Constructor for the `BioeconomicSim` class
#'
#' A constructor for the `BioeconomicSim` class. This need not be called directly, as this class is generated automatically via the `project()` function.
#'@param params An object of class `BioeconomicParams`.
#'@param R The recruitment across the simulation. This is either a `numeric` giving the recruitment at each time step, or a function of the previous years adult stock abundance.
#'@param t_start An `integer` specifying the start time for the simulation. The default is `1`
#'@param t_end An `integer` specifying the end time for the simulation. The default is `1 + length(R)`
#'@param R_init A `numeric` giving the initial recruitment value. Only required if `R` is a function of stock size.
#'@return An object of type \linkS4class{BioeconomicSim}
#'@examples
#'data("seabass")
#'recruitment_fun <- function(stock){
#'  return(
#'        1.492*stock/(7.2e-4 + 1.87e-07*stock)*exp(rnorm(1,mean=0,sd=0.9))
#'        )
#'}
#'sim <- BioeconomicSim(params, recruitment_fun, t_start = 1980, t_end = 2020, R_init = 1e5)
#'@export
BioeconomicSim <- function(params, R, t_start, t_end, R_init = NULL) {

  t_max = t_end - t_start + 1
  states <- init_states(params, R, R_init, t_max)

  rec_func <- NULL
  if(is.function(R))
    rec_func <- R

  sim <-new('BioeconomicSim',
            params = params,
            rec_func = rec_func,
            t_start = t_start,
            t_end = t_end,
            states = states)

  return(sim)
}

setClassUnion("funcorNULL", members = c("function", "NULL"))

#### Class definition ####
#' A class to hold the results of a simulation.
#'
#' @slot params A `BioeconomicParams` object encoding the parameters for the model.
#' @slot rec_func The recruitment function
#' @slot t_start The start time of the simulation.
#' @slot t_end The end time of the simulation.
#' @slot states A `data.frame` containing the states of the system for each time step of the simulation.
#' @seealso \link{BioeconomicSim}
#' @details
#' The states of the simulation at each time step are stored in the `states` slot of this object. The `states` slot contains a \eqn{T \times 17} matrix where \eqn{T} is the duration of the simulation. Each column represents a different model output and each row gives the value of those outputs at a time step. The columns in this matrix are:
#' 1. R - The recruitment (numbers)
#' 2. SJ - The size of the juvenile stock (numbers)
#' 3. SA - The size of the adult stock (numbers)
#' 4. TC - The number of commercial trips
#' 5. TR - The number of recreationl trips.
#' 6. EC - The commercial effort
#' 7. ER - The recreational effort
#' 8. FC - The commercial fishing mortality
#' 9. FR - The recreational fishing mortality
#' 10. tau - The profit of the commercial fleet
#' 11. K - The catch per unit effort of the recreational fleet
#' 12. LR - The landed catch of the recreational fleet (tonnes)
#' 13. LC - The landed catch of the commercial fleet (tonnes)
#' 14. D - The proportion of the juvenile stock that mature into the adult stock
#' 15. VR - The GVA of the recreational fleet
#' 16. VC - The GVA of the commercial fleet
#' 17. ZA - The total mortality on the adult stock
#' @export
setClass(
  "BioeconomicSim",
  slots = c(params = "BioeconomicParams",
            rec_func = "funcorNULL",
            t_start = "numeric",
            t_end = "numeric",
            states = "data.frame")
)

#' Summary plot for `BioeconomicSim` objects.
#'
#' After running a simulation, produces 6 plots: stock size, activity (commercial fleet), activity (recreational fleet), economic impact (total across both fleets), economic impact (commercial fleet), economic impact (recreational fleet).
#' @examples
#' #loads in the relevant parameters
#' data("seabass")
#' # Runs the simulation with 20 years of high recruitment followed by 20 years of low recruitment
#' R <- c(rep(23151200, 20), rep(757600, 20))
#' sim <- project(params, R = R)
#' plot(sim)
#'@export
plot.BioeconomicSim <- function(sim){
  cowplot::plot_grid(plotStock(sim, stock = "B"),
                     plotActivity(sim, fleet = "C"),
                     plotActivity(sim, fleet = "R"),
                     plotEconomicImpact(sim, fleet = "T"),
                     plotEconomicImpact(sim, fleet = "C"),
                     plotEconomicImpact(sim, fleet = "R"),
                     nrow = 6, ncol = 1)
}




init_states <- function(params, R, R_init, t_max){

  state_names <- c("R", "SJ", "SA", "TC", "TR", "EC", "ER", "FC", "FR", "tau", "K", "LR", "LC", "D", "VR", "VC", "ZA")
  states <- matrix(NA, t_max, length(state_names))
  colnames(states) <- state_names
  states <- as.data.frame(states)

  if(!is.function(R))
    states$R <- R
  else
    states$R[1] <- R_init

  # Initialisation
  states$SJ[1] = params@initialSJ
  states$SA[1] = params@initialSA
  states$TC[1] = params@initialTC
  states$TR[1] = params@initialTR

  states$EC[1] = params@chiC * states$TC[1]
  states$ER[1] = params@chiR * states$TR[1]
  states$FC[1] = params@qC * states$EC[1]
  states$FR[1] = params@qR * states$ER[1]

  states$ZA[1] = params@muA + (1-params@varphi * params@delta)*states$FR[1] +  (1 - params@Gamma * params@eta)*states$FC[1]

  states$D[1] = exp(-6*params@muJ)

  states$LC[1] <- params@WC*(1 - params@eta)  *states$FC[1]*(1 - exp(-states$ZA[1]))*states$SA[1] / states$ZA[1]
  states$LR[1] <- params@WR*(1 - params@delta)*states$FR[1]*(1 - exp(-states$ZA[1]))*states$SA[1] / states$ZA[1]


  states$VC[1] = params@sigma*params@nu*states$LC[1]
  states$VR[1] = params@zeta*params@lambda*states$TR[1]
  states$tau[1] = params@nu*states$LC[1] - (params@baromega*states$EC[1] + params@Lambda*params@phi)
  states$K[1] = states$FR[1]*states$SA[1]/states$ER[1] #params@b+1
  return(states)

}







R/BioeconomicParams-class.R
#' Constructor for the `BioeconomicParams` class
#'
#' A constructor for the `BioeconomicParams` class.
#'@param file A csv file contatining the input parameters and their values
#'@return An object of type \code{\linkS4class{BioeconomicParams}}
#'@details
#'The `file` params csv file should contain all parameter values required to run the model (see the \code{\linkS4class{BioeconomicParams}} class definition for a full list of required parameters), the first column containing the parameter names and the second column containing the corresponding values.
#'@export
#'@examples
#'\dontrun{
#'params <- BioeconomicParams(file = "params.csv")
#'}
BioeconomicParams <- function(file) {

  params_raw <- read.csv(file)[, 1:2]
  params_raw <- split(params_raw[, 2], params_raw[, 1])

  bioeceonomic_params <- do.call(function(...) new('BioeconomicParams', ...), params_raw)

  return(bioeceonomic_params)
}

#### Class definition ####
#' A class for bioeconomic parameters.
#' @slot g The magnitude of commercial trip change per year.
#' @slot p The magnitude of recreational trip change per year.
#' @slot b Recreational catch per effort threshold.
#' @slot qC Commercial catchability per effort.
#' @slot qR Recreational catchability per effort.
#' @slot muA The natural mortality of the adult stock.
#' @slot muJ The natural mortality of the juvenile stock.
#' @slot chiR Effort per recreational trip.
#' @slot chiC Days per commercial trip.
#' @slot theta Max number of commercial trips per year.
#' @slot gamma Max number of recreational trips per year.
#' @slot delta Proportion of recreational catch released.
#' @slot varphi Proportion of released recreational catch that survives.
#' @slot eta Commercial discard rate.
#' @slot Gamma Proportion of commercial discards that survive.
#' @slot WR Average weight of recreationally landed fish (tonnes).
#' @slot WC Average weight of commercial landed fish (tonnes).
#' @slot nu Commercial fish price per tonnes.
#' @slot baromega Variable cost per commercial day at sea.
#' @slot Lambda Proportion fixed cost attributable to sea bass.
#' @slot phi Fixed cost per fleet per year.
#' @slot sigma GVA multiplier for commercial Revenue.
#' @slot lambda Average per trip expenditure of recreational anglers.
#' @slot zeta GVA multiplier for recreational trip expenditure.
#' @slot initialSJ The starting value of the size of the juvenile stock (in numbers).
#' @slot initialSA The starting value of the size of the adult stock (in numbers).
#' @slot initialTC The starting value for the number of commercial trips.
#' @slot initialTR The starting value for the number of recreational trips.
#' @export
setClass(
  "BioeconomicParams",
  slots = c(theta = "numeric",
             gamma = "numeric",
             g = "numeric",
             p = "numeric",
             b = "numeric",
             WR = "numeric",
             WC = "numeric",
             phiC = "numeric",
             phiR = "numeric",
             muA = "numeric",
             muJ = "numeric",
             sigma = "numeric",
             zeta = "numeric",
             nu = "numeric",
             baromega = "numeric",
             phi = "numeric",
             lambda = "numeric",
             chiR = "numeric",
             chiC = "numeric",
             delta = "numeric",
             eta = "numeric",
             Lambda = "numeric",
             initialSJ = "numeric",
             initialSA = "numeric",
             initialTC = "numeric",
             initialTR = "numeric",
             varphi = "numeric",
             Gamma = "numeric",
            qC = "numeric",
            qR = "numeric")
)


R/simulation_functions.R
step = function(sim, t){

  params <- sim@params

  states <- sim@states[t, ]
  previous_states <- sim@states[t - 1, ]


  if(sum(is.na(previous_states)) != 0){
    stop("Invalid states detected.")
  }

  #Step 1. Work out the number of trips that should occur
  #Commercial trips
  previous_profit <- previous_states$tau
  if(t > 2){
    previous_profit <- sim@states[t - 2, ]$tau
  }

  if (previous_profit >= 0 && previous_states$TC < params@theta){
    states$TC <- min((1 + params@g)*previous_states$TC, params@theta)
  }else if(previous_profit < 0){
    states$TC <- (1 - params@g)*previous_states$TC
  }else{
    states$TC <- params@theta
  }

  #Recreational trips
  if(previous_states$K >= params@b && previous_states$TR < params@gamma){
    states$TR <- min((1 + params@p)*previous_states$TR, params@gamma)
  }else if(previous_states$K < params@b){
    states$TR <- (1 - params@p)*previous_states$TR
  }else{
    states$TR <- params@gamma
  }

  #Step 2: Translate trips into effort and fishing pressure
  states$EC <- params@chiC * states$TC
  states$ER <- params@chiR * states$TR
  states$FC <- params@qC * states$EC
  states$FR <- params@qR * states$ER

  #Step 3: Stock effects: Recruitment
  if(is.function(sim@rec_func)){
    states$R <- sim@rec_func(previous_states$SA)
  }

  #Step 4: Stock effects: Stock sizes
  states$SJ = previous_states$SJ + states$R - (1 - exp(-params@muJ))*previous_states$SJ - previous_states$D*previous_states$SJ
  states$ZA <- params@muA + (1-params@varphi * params@delta)*states$FR +  (1 - params@Gamma * params@eta)*states$FC
  states$SA = exp(-states$ZA)*previous_states$SA + previous_states$D*previous_states$SJ

  states$D <- exp(-6*params@muJ)
  if(t > 5){
    states$D <- exp(-6*params@muJ) * sim@states[t - 5, "R"]/ states$SJ
  }

  #Step 5: Economic impacts: Commercial fleet
  states$LC  <- params@WC * (1-params@eta) * states$FC * (1 - exp(-states$ZA)) * states$SA / states$ZA

  states$tau <- params@nu * states$LC - (params@baromega * states$EC + params@Lambda * params@phi)
  states$VC  <- params@sigma * params@nu * states$LC

  #Step 6: Economic impacts: Recreational fleet
  states$LR <- params@WR * (1 - params@delta) * states$FR * (1 - exp(-states$ZA)) * states$SA / states$ZA
  states$VR <- params@zeta * params@lambda * states$TR


  #Step 7: CPUE for recreational fleet.
  states$K <- (1-exp(-states$FR))*states$SA/states$ER

  return(states)
}

#' Project a FishRAM simulation.
#'
#' Runs the FishRAM model for a specified recruitment scenario. Full details of the model specification can be found in the supporting information of Tidbury et al. (2021), however some details have been extended for this package (see details below).
#' @param params An object of type `BioeconomicParams` encoding the model parameters.
#' @param R A `numeric` of `function` specifying the recruitment of the stock across the simulation. See details
#' @param t_start The year which forms the start of the simulation. The default is 1.
#' @param t_end The year which forms the end of the simulation. The default is `length(R) + t_start - 1`
#' @param R_init The initial recruitment value. Only required if `R` is a function.
#' @return An object of class [BioeconomicSim-class]. See the `BioeconomicSim-class` help file for a full list of model outputs included in this object.
#' @details
#' Recruitment can be specified in 3 ways using different forms of the `R` parameter:
#' * A `numeric` of length 1, giving the constant recruitment across the duration of the simulation.
#' * A `numeric` of length `t_end - t_start + 1`, specifying the recruitment at each time step of the simulation.
#' * A `function` which calculates recruitment from the adult stock size at each time step. This function should have exactly one argument (the adult stock size) and return a `numeric` of length 1. If using a function, the `R_init` argument should also be passed through to specify the initial recruitment at the start of the simulation.
#'
#'
#' @section Model extensions:
#' *Mortality reparametrisation.* To allow for a larger range of mortalities, the mortality in `FishRAM` is parametrised differently to the mortality in Tidbury et al. (2021). The stock dynamics for the adult population are instead calculated through
#' \deqn{S_{A, t} = \exp(M_A + (1 - \varphi \delta)F_{R, t} + (1-\Gamma \eta)F_{C, t})S_{A, t-1} + D_{t}S_{J, t}}.
#' With the landings of each fleet now calculated as
#' \deqn{L_{R, t} = W_C \frac{(1-\delta)F_{R, t}}{M_A + (1 - \varphi \delta)F_{R, t} + (1-\Gamma \eta)F_{C, t}}S_{A, t}}
#' \deqn{L_{C, t} = W_C \frac{(1-\eta)F_{C, t}}{M_A + (1 - \varphi \delta)F_{R, t} + (1-\Gamma \eta)F_{C, t}}S_{A, t}}
#' This is equivalent to assuming that all mortalities (commercial, recreational, and natural) are applied simultaneously to the adult stock.
#'
#' *Initial values*
#' For the first 5 years of the simulation, some quantities relying on the history of simulation are calculated differently, by assuming some values were static prior to the start of the simulation.
#' * Number of trips. The number of trips of the commercial fleet \eqn{T_C} depends on the profit of the fleet from 2 years ago. The commercial fleet profit is therefore assumed to be constant before the simulation starts (i.e. for time \eqn{t <= 0}) to ensure that \eqn{T_C} can be calculated for the first 2 years.
#' * Maturation. The fraction of juveniles maturing into adults in a given year \eqn{D_t} is estimated from the recruitment values from 5 years previously. Recruitment is therefore assumed to be constant before the start of the simulation so that $D_t$ can be calculated for the first 5 years.
#'
#' @references Hannah J Tidbury, Angela Muench, Philip D Lamb, Kieran Hyder, Balancing biological and economic goals in commercial and recreational fisheries: systems modelling of sea bass fisheries, *ICES Journal of Marine Science*, Volume 78, Issue 5, August 2021, Pages 1793-1803
#' @export
#' @examples
#' #loads in the relevant parameters
#' data("seabass")
#' sim <- project(params, R = 8606000, t_start = 1990, t_end = 2030)
#'
#' #Runs the simulation with 20 years of high recruitment followed
#' #by 20 years of low recruitment
#' R <- c(rep(23151200, 20), rep(757600, 20))
#' sim <- project(params, R = R)
#'
#' #Generates recruitment from a type II functional response
#' #recruitment function with log Gaussian noise.
#' rec_func <- function(stock){
#'   return(
#'     1.5@*stock/(7e-4 + 1.3e-4 @* 1.5e-3 @* stock)@*exp(rnorm(1,mean=0,sd=0.9))
#'   )
#' }
#' sim <- project(params, R = rec_func, t_start = 2010, t_end = 2050, R_init = 1e4)
project = function(params, R, t_start = 1, t_end = (length(R) + t_start - 1), R_init = NULL){
  t_max <- t_end - t_start + 1

  if(!is(params, "BioeconomicParams")){
    stop("params should be a BioeconomicParams object")
  }
  if(!(is.numeric(R) || is.function(R))){
    stop("R should be a numeric vector or function specifying recruitment")
  }
 if(!(length(R) == 1 || length(R) == t_max)){
  stop("R should have length 1 or length equal to the duration of the simulation")
 }

  if(is.function(R) && is.null(R_init)){
    stop("R is a function, but no initial values have been specified. The initial value should be specified using the R_init parameter.")
  }

  #TODO: Think about chaining sim together, so setting your own initial values outside of the params object?
  sim <- BioeconomicSim(params, R, t_start, t_end, R_init)

  for (t in 2:t_max){
    sim@states[t, ] = step(sim, t)
  }

  return(sim)
}
